File: substring_with_concatenation_of_all_words.java
import java.util.*;

class Solution {
    public List<Integer> findSubstring(String s, String[] words) {
        List<Integer> res = new ArrayList<>();
        
        if (s == null || words == null || words.length == 0) {
            return res;
        }
        
        Map<String, Integer> wordFreq = new HashMap<>();
        for (String word : words) {
            wordFreq.put(word, 1 + wordFreq.getOrDefault(word, 0));
        }
        
        int wordLen = words[0].length();
        int windowLen = words.length * wordLen;
        
        for (int i = 0; i <= s.length() - windowLen; i++) {
            Map<String, Integer> substrFreq = new HashMap<>();
            int j = i;
            
            while (j < i + windowLen) {
                String curWord = s.substring(j, j + wordLen);
                
                if (!wordFreq.containsKey(curWord)) {
                    break;
                }
                
                substrFreq.put(curWord, 1 + substrFreq.getOrDefault(curWord, 0));
                
                if (substrFreq.get(curWord) > wordFreq.get(curWord)) {
                    break;
                }
                
                j += wordLen;
            }
            
            if (j == i + windowLen) {
                res.add(i);
            }
        }
        
        return res;
    }
}

File: substring_with_concatenation_of_all_words.py
class Solution:
    def findSubstring(self, s: str, words: List[str]) -> List[int]:

        if not s or not words:
            return []
        
        word_freq = {}
        for word in words:
            word_freq[word] = 1 + word_freq.get(word, 0)
        
        word_len = len(words[0])
        window_len = len(words) * word_len

        res = []

        for i in range(len(s) - window_len + 1):
            substr_freq = {}
            j = i

            while j < i + window_len:
                cur_word = s[j:j+word_len]

                if cur_word not in word_freq:
                    break
                substr_freq[cur_word] = substr_freq.get(cur_word, 0) + 1

                if substr_freq[cur_word] > word_freq[cur_word]:
                    break
                j += word_len
            
            if j == i + window_len:
                res.append(i)
        
        return res




        
        
